src/dbnode/encoding/m3tsz/iterator.go

```diff
@@ -384,30 +227,353 @@ func (it *readerIterator) hasNext() bool {
 func (it *readerIterator) Reset(reader io.Reader) {
 	it.is.Reset(reader)
-	it.t = time.Time{}
-	it.dt = 0
+	it.tsIterator = NewTimestampIterator(it.opts)
 	it.vb = 0
 	it.xor = 0
-	it.done = false
 	it.err = nil
-	it.ant = nil
 	it.isFloat = false
 	it.intVal = 0.0
 	it.mult = 0
 	it.sig = 0
-	it.tu = xtime.None
 	it.closed = false
 }
 
+// Close closes the ReaderIterator.
 func (it *readerIterator) Close() {
 	if it.closed {
 		return
 	}
+
 	it.closed = true
 	pool := it.opts.ReaderIteratorPool()
 	if pool != nil {
 		pool.Put(it)
 	}
 }
+
+// TimestampIterator encapsulates all the state required for iterator over
+// delta-of-delta compresed timestamps.
+type TimestampIterator struct {
+	PrevTime      time.Time
+	PrevTimeDelta time.Duration
+	PrevAnt       ts.Annotation
+
+	TimeUnit        xtime.Unit
+	TimeUnitChanged bool
+
+	Done bool
+
+	Opts encoding.Options
+}
+
+// NewTimestampIterator creates a new TimestampIterator.
+func NewTimestampIterator(opts encoding.Options) TimestampIterator {
+	return TimestampIterator{
+		Opts: opts,
+	}
+}
+
+// ReadTimestamp reads the first or next timestamp.
+func (it *TimestampIterator) ReadTimestamp(stream encoding.IStream) (bool, bool, error) {
+	it.PrevAnt = nil
+	it.TimeUnitChanged = false
+
+	var (
+		first = false
+		err   error
+	)
+	if it.PrevTime.IsZero() {
+		first = true
+		err = it.readFirstTimestamp(stream)
+	} else {
+		err = it.readNextTimestamp(stream)
+	}
+	if err != nil {
+		return false, false, err
+	}
+
+	// NB(xichen): reset time delta to 0 when there is a time unit change to be
+	// consistent with the encoder.
+	if it.TimeUnitChanged {
+		it.PrevTimeDelta = 0
+	}
+
+	return first, it.Done, nil
+}
+
+func (it *TimestampIterator) readFirstTimestamp(stream encoding.IStream) error {
+	ntBits, err := stream.ReadBits(64)
+	if err != nil {
+		return err
+	}
+
+	nt := int64(ntBits)
+	// NB(xichen): first time stamp is always normalized to nanoseconds.
+	st := xtime.FromNormalizedTime(nt, time.Nanosecond)
+	it.TimeUnit = initialTimeUnit(st, it.Opts.DefaultTimeUnit())
+
+	err = it.readNextTimestamp(stream)
+	if err != nil {
+		return err
+	}
+
+	it.PrevTime = st.Add(it.PrevTimeDelta)
+	return nil
+}
+
+func (it *TimestampIterator) readNextTimestamp(stream encoding.IStream) error {
+	dod, err := it.readMarkerOrDeltaOfDelta(stream)
+	if err != nil {
+		return err
+	}
+
+	it.PrevTimeDelta += dod
+	it.PrevTime = it.PrevTime.Add(it.PrevTimeDelta)
+	return nil
+}
+
+func (it *TimestampIterator) tryReadMarker(stream encoding.IStream) (time.Duration, bool, error) {
+	mes := it.Opts.MarkerEncodingScheme()
+	numBits := mes.NumOpcodeBits() + mes.NumValueBits()
+	opcodeAndValue, success := it.tryPeekBits(stream, numBits)
+	if !success {
+		return 0, false, nil
+	}
+
+	opcode := opcodeAndValue >> uint(mes.NumValueBits())
+	if opcode != mes.Opcode() {
+		return 0, false, nil
+	}
+
+	var (
+		valueMask   = (1 << uint(mes.NumValueBits())) - 1
+		markerValue = int64(opcodeAndValue & uint64(valueMask))
```

Comment:
You need to be careful here that the possible combination of your bits for your proto encoding does not conflict with potential markers for EOS, TimeUnit or Annotation. For the M3TSZ encoder int component, we picked the ordering of the opCode bits such that a conflict was not possible, otherwise, you won't be able to read what you've encoded correctly. This was also done by the TSZ XOR encoder originally.

Reply:
Yeah I'm not actually sure how to resolve this without just re-implementing all the logic myself :/

Reply:
One "relatively" straightforward solution would be to pass a flag to the `TimestampIterator` to tell it not to do the marker peeking logic since I dont need the `EndOfStream` marker or the `Annotation` marker anways. That just leaves the time unit changed marker stuff which I can incorporate myself I think

Reply:
Yeah probably the way to go for Proto and maybe other encoders. 

It's going to be increasingly harder to ensure no conflicts with more encoders. Most of the existing markers consist of multiple 0's and 1's in a row (4 minimum I believe?), so we used that to ensure not many header bits overlaps. If you have lots of header bits, might be worth just using 2-3 bits instead of 1 bit per header to have more combinations. Also fine to just disable this logic for the proto encoder if it's too complicated.

Reply:
Yeah that sounds good, re: approach of not doing peek logic for proto encoder.